src/libfuncs/pedersen.rs

```diff
@@ -1,3 +1,272 @@
 //! # Pedersen hashing libfuncs
 //!
-//! TODO
+
+use super::{LibfuncBuilder, LibfuncHelper};
+use crate::{
+    error::{
+        libfuncs::{Error, Result},
+        CoreTypeBuilderError,
+    },
+    metadata::{runtime_bindings::RuntimeBindingsMeta, MetadataStorage},
+    types::TypeBuilder,
+    utils::get_integer_layout,
+};
+use cairo_lang_sierra::{
+    extensions::{
+        lib_func::SignatureOnlyConcreteLibfunc, pedersen::PedersenConcreteLibfunc, ConcreteLibfunc,
+        GenericLibfunc, GenericType,
+    },
+    program_registry::ProgramRegistry,
+};
+use melior::{
+    dialect::{
+        arith,
+        llvm::{self, LoadStoreOptions},
+    },
+    ir::{
+        attribute::{IntegerAttribute, StringAttribute},
+        operation::OperationBuilder,
+        r#type::IntegerType,
+        Block, Identifier, Location,
+    },
+    Context,
+};
+
+/// Select and call the correct libfunc builder function from the selector.
+pub fn build<'ctx, 'this, TType, TLibfunc>(
+    context: &'ctx Context,
+    registry: &ProgramRegistry<TType, TLibfunc>,
+    entry: &'this Block<'ctx>,
+    location: Location<'ctx>,
+    helper: &LibfuncHelper<'ctx, 'this>,
+    metadata: &mut MetadataStorage,
+    selector: &PedersenConcreteLibfunc,
+) -> Result<()>
+where
+    TType: GenericType,
+    TLibfunc: GenericLibfunc,
+    <TType as GenericType>::Concrete: TypeBuilder<TType, TLibfunc, Error = CoreTypeBuilderError>,
+    <TLibfunc as GenericLibfunc>::Concrete: LibfuncBuilder<TType, TLibfunc, Error = Error>,
+{
+    match selector {
+        PedersenConcreteLibfunc::PedersenHash(info) => {
+            build_pedersen(context, registry, entry, location, helper, metadata, info)
+        }
+    }
+}
+
+pub fn build_pedersen<'ctx, TType, TLibfunc>(
+    context: &'ctx Context,
+    registry: &ProgramRegistry<TType, TLibfunc>,
+    entry: &'ctx Block<'ctx>,
+    location: Location<'ctx>,
+    helper: &LibfuncHelper<'ctx, '_>,
+    metadata: &mut MetadataStorage,
+    info: &SignatureOnlyConcreteLibfunc,
+) -> Result<()>
+where
+    TType: GenericType,
+    TLibfunc: GenericLibfunc,
+    <TType as GenericType>::Concrete: TypeBuilder<TType, TLibfunc, Error = CoreTypeBuilderError>,
+    <TLibfunc as GenericLibfunc>::Concrete: LibfuncBuilder<TType, TLibfunc, Error = Error>,
+{
+    {
+        metadata
+            .get_mut::<RuntimeBindingsMeta>()
+            .expect("Runtime library not available.");
+    }
```

Comment:
Is there any reason why you have it within a block?

Reply:
yes, lifetime issues, metadata is borrowed mutable somewhere

Reply:
ok never mind the issue is no longer there